### PR TITLE
#308: Equality

### PR DESCRIPTION
**Pull request by fabriciofx**
As per #308.

### CODE REVIEW

File: src/main/java/org/cactoos/func/Equality.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            if (idx >= rght.length) {
+                result = 1;
+                break;
+            }
+            result = lft[idx] - rght[idx];
+            if (result != 0) {
+                break;
+            }
+        }
+        return (int) Math.signum(result);
```
Comment (yegor256) [inline src/main/java/org/cactoos/func/Equality.java]: @fabriciofx please, check the article on the blog, the code is changed there. Pay attention to this thread of comments: http://www.yegor256.com/2017/07/11/how-to-redesign-equals.html#comment-3420679770

File: src/test/java/org/cactoos/func/EqualityTest.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * Test case for {@link Equality}.
+ *
+ * @author Yegor Bugayenko (yegor256@gmail.com)
+ * @author Fabricio Cabral (fabriciofx@gmail.com)
+ * @version $Id$
+ * @since 0.12
+ * @checkstyle JavadocMethodCheck (500 lines)
+ */
+public final class EqualityTest {
```
Comment (yegor256) [inline src/test/java/org/cactoos/func/EqualityTest.java]: @fabriciofx you may be interested in more tests, as we did in this thread: http://www.yegor256.com/2017/07/11/how-to-redesign-equals.html#comment-3420679770

**Comment by yegor256** [comment]
@fabriciofx thanks for the code, looks good, but please pay attention to my comments above

File: src/main/java/org/cactoos/func/Equality.java (hunk 1 above)
Comment (fabriciofx) [inline src/main/java/org/cactoos/func/Equality.java]: @yegor256 Right. I made some changes. Could you check them?

File: src/test/java/org/cactoos/func/EqualityTest.java (hunk 2 above)
Comment (fabriciofx) [inline src/test/java/org/cactoos/func/EqualityTest.java]: @yegor256 Right. I made some changes. Could you check them?

**Comment by llorllale** [comment]
@0crat in

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/yegor256/cactoos/issues/309#issuecomment-382077422))

@llorllale Job [#309](https://github.com/yegor256/cactoos/issues/309) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#309](https://github.com/yegor256/cactoos/issues/309) is assigned to @neonailol[/z](https://www.0crat.com/u/neonailol), here is [why](https://www.0crat.com/footprint/C63314D6Z/1071737473); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @llorllale[/z](https://www.0crat.com/u/llorllale) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer

**Comment by 0crat** [comment]
@neonailol[/z](https://www.0crat.com/u/neonailol) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by llorllale** [comment]
@0crat resign @neonailol

**Comment by 0crat** [comment]
> @0crat resign @neonailol  ([here](https://github.com/yegor256/cactoos/issues/309#issuecomment-384301372))

@llorllale The user @neonailol[/z](https://www.0crat.com/u/neonailol) resigned from [#309](https://github.com/yegor256/cactoos/issues/309), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
This pull request [#309](https://github.com/yegor256/cactoos/issues/309) is assigned to @sharpler[/z](https://www.0crat.com/u/sharpler), here is [why](https://www.0crat.com/footprint/C63314D6Z/1151438088); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @llorllale[/z](https://www.0crat.com/u/llorllale) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer

File: src/main/java/org/cactoos/func/Equality.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    public Integer value() throws Exception {
+        final byte[] left = this.lft.asBytes();
+        final byte[] right = this.rght.asBytes();
+        final int max = Math.max(left.length, right.length);
+        int result = 0;
+        for (int idx = max; idx > 0; --idx) {
+            byte first = 0;
+            if (idx <= left.length) {
```
Comment (SharplEr) [inline src/main/java/org/cactoos/func/Equality.java]: @fabriciofx Looks like code can be more optimal. You can remove `if` statements from the loop. Get a longer array and check its last bytes is not zero. After that, you can compare a shorter array with remaining part of the longer one with only one `if` statement.

**Comment by SharplEr** [comment]
@fabriciofx Please pay attention to my remarks.

**Comment by SharplEr** [comment]
@fabriciofx ping)

**Comment by 0crat** [comment]
@sharpler[/z](https://www.0crat.com/u/sharpler) this job was assigned to you 7days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by Vatavuk** [comment]
@llorllale @SharplEr Fabricio is inactive. This job is assigned to me and I've created new PR #809

**Comment by 0crat** [comment]
The user @sharpler[/z](https://www.0crat.com/u/sharpler) resigned from [#309](https://github.com/yegor256/cactoos/issues/309), please stop working. Reason for job resignation: It is older than 10 days, see [§8](http://www.zerocracy.com/policy.html#8)

**Comment by 0crat** [comment]
Resigned on delay, see [§8](http://www.zerocracy.com/policy.html#8): -15 point(s) just awarded to @sharpler[/z](https://www.0crat.com/u/sharpler)

**Comment by llorllale** [comment]
@proshin-roman thanks

**Comment by elenavolokhova** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @proshin-roman[/z](https://www.0crat.com/u/proshin-roman)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova)

**Pull request closed without merging**